pkg/reconciler/buildrun/resources/params.go

```diff
@@ -53,3 +65,511 @@ func overrideParams(originalParams []buildv1alpha1.ParamValue, overrideParams []
 func IsSystemReservedParameter(param string) bool {
 	return systemReservedParamKeys[param] || strings.HasPrefix(param, "shp-")
 }
+
+// FindParameterByName returns the first entry in a Parameter array with a specified name, or nil
+func FindParameterByName(parameters []buildv1alpha1.Parameter, name string) *buildv1alpha1.Parameter {
+	for _, candidate := range parameters {
+		if candidate.Name == name {
+			return &candidate
+		}
+	}
+
+	return nil
+}
+
+// FindParamValueByName returns the first entry in a ParamValue array with a specified name, or nil
+func FindParamValueByName(paramValues []buildv1alpha1.ParamValue, name string) *buildv1alpha1.ParamValue {
+	for _, candidate := range paramValues {
+		if candidate.Name == name {
+			return &candidate
+		}
+	}
+
+	return nil
+}
+
+// HandleTaskRunParam makes the necessary changes to a TaskRun for a parameter
+func HandleTaskRunParam(taskRun *pipeline.TaskRun, parameterDefinition *buildv1alpha1.Parameter, paramValue buildv1alpha1.ParamValue) error {
+	taskRunParam := pipeline.Param{
+		Name:  paramValue.Name,
+		Value: pipeline.ArrayOrString{},
+	}
+
+	switch parameterDefinition.Type {
+	case "": // string is default
+		fallthrough
+	case buildv1alpha1.ParameterTypeString:
+		taskRunParam.Value.Type = pipeline.ParamTypeString
+
+		if paramValue.SingleValue != nil {
+			if paramValue.SingleValue.ConfigMapValue != nil {
+				envVarName, err := addConfigMapEnvVar(taskRun, paramValue.Name, paramValue.SingleValue.ConfigMapValue.Name, paramValue.SingleValue.ConfigMapValue.Key)
+
+				if err != nil {
+					return err
+				}
+
+				envVarExpression := fmt.Sprintf("$(%s)", envVarName)
+				if paramValue.SingleValue.ConfigMapValue.Format != nil {
+					taskRunParam.Value.StringVal = strings.ReplaceAll(*paramValue.SingleValue.ConfigMapValue.Format, "${CONFIGMAP_VALUE}", envVarExpression)
+				} else {
+					taskRunParam.Value.StringVal = envVarExpression
+				}
+			} else if paramValue.SingleValue.SecretValue != nil {
+				envVarName, err := addSecretEnvVar(taskRun, paramValue.Name, paramValue.SingleValue.SecretValue.Name, paramValue.SingleValue.SecretValue.Key)
+				if err != nil {
+					return err
+				}
+
+				envVarExpression := fmt.Sprintf("$(%s)", envVarName)
+				if paramValue.SingleValue.SecretValue.Format != nil {
+					taskRunParam.Value.StringVal = strings.ReplaceAll(*paramValue.SingleValue.SecretValue.Format, "${SECRET_VALUE}", envVarExpression)
+				} else {
+					taskRunParam.Value.StringVal = envVarExpression
+				}
+			} else if paramValue.SingleValue.Value != nil {
```

Comment:
I suspect that if you converted all these if - then - else - if's to more switch / case's it would improve readability, maybe reduce some sprawl.

WDYT?

Reply:
Done.

Reply:
cool thanks - it reads more easily now IMO